File: src/content/cre/getting-started/snippets/part-2-main.ts
import {
  CronCapability,
  HTTPClient,
  handler,
  consensusMedianAggregation,
  Runner,
  type NodeRuntime,
  type Runtime,
} from "@chainlink/cre-sdk"

type Config = {
  schedule: string
  apiUrl: string
}

type MyResult = {
  result: bigint
}

const initWorkflow = (config: Config) => {
  const cron = new CronCapability()

  return [handler(cron.trigger({ schedule: config.schedule }), onCronTrigger)]
}

// fetchMathResult is the function passed to the runInNodeMode helper.
// It contains the logic for making the request and parsing the response.
const fetchMathResult = (nodeRuntime: NodeRuntime<Config>): bigint => {
  const httpClient = new HTTPClient()

  const req = {
    url: nodeRuntime.config.apiUrl,
    method: "GET" as const,
  }

  // Send the request using the HTTP client
  const resp = httpClient.sendRequest(nodeRuntime, req).result()

  // The mathjs.org API returns the result as a raw string in the body.
  // We need to parse it into a bigint.
  const bodyText = new TextDecoder().decode(resp.body)
  const val = BigInt(bodyText.trim())

  return val
}

const onCronTrigger = (runtime: Runtime<Config>): MyResult => {
  runtime.log("Hello, Calculator! Workflow triggered.")
  // Use runInNodeMode to execute the offchain fetch.
  // The API returns a random number, so each node can get a different result.
  // We use median consensus to find a single, trusted value.
  const result = runtime.runInNodeMode(fetchMathResult, consensusMedianAggregation())().result()

  runtime.log(`Successfully fetched and aggregated math result: ${result}`)

  return {
    result,
  }
}

export async function main() {
  const runner = await Runner.newRunner<Config>()
  await runner.run(initWorkflow)
}

File: src/content/cre/getting-started/snippets/part-2-main.go
//go:build wasip1

package main

import (
	"fmt"
	"log/slog"
	"math/big"

	"github.com/smartcontractkit/cre-sdk-go/capabilities/networking/http"
	"github.com/smartcontractkit/cre-sdk-go/capabilities/scheduler/cron"
	"github.com/smartcontractkit/cre-sdk-go/cre"
	"github.com/smartcontractkit/cre-sdk-go/cre/wasm"
)

// Add the ApiUrl to your config struct
type Config struct {
	Schedule string `json:"schedule"`
	ApiUrl   string `json:"apiUrl"`
}

type MyResult struct {
	Result *big.Int
}

func InitWorkflow(config *Config, logger *slog.Logger, secretsProvider cre.SecretsProvider) (cre.Workflow[*Config], error) {
	return cre.Workflow[*Config]{
		cre.Handler(
			cron.Trigger(&cron.Config{Schedule: config.Schedule}),
			onCronTrigger,
		),
	}, nil
}

// fetchMathResult is the function passed to the http.SendRequest helper.
// It contains the logic for making the request and parsing the response.
func fetchMathResult(config *Config, logger *slog.Logger, sendRequester *http.SendRequester) (*big.Int, error) {
	req := &http.Request{
		Url:    config.ApiUrl,
		Method: "GET",
	}

	// Send the request using the provided sendRequester
	resp, err := sendRequester.SendRequest(req).Await()
	if err != nil {
		return nil, fmt.Errorf("failed to get API response: %w", err)
	}

	// The mathjs.org API returns the result as a raw string in the body.
	// We need to parse it into a big.Int.
	val, ok := new(big.Int).SetString(string(resp.Body), 10)
	if !ok {
		return nil, fmt.Errorf("failed to parse API response into big.Int")
	}
	return val, nil
}

// onCronTrigger is our main DON-level callback.
func onCronTrigger(config *Config, runtime cre.Runtime, trigger *cron.Payload) (*MyResult, error) {
	logger := runtime.Logger()
	logger.Info("Hello, Calculator! Workflow triggered.")

	client := &http.Client{}
	// Use the http.SendRequest helper to execute the offchain fetch.
	mathPromise := http.SendRequest(config, runtime, client,
		fetchMathResult,
		// The API returns a random number, so each node can get a different result. We use Median Aggregation to find a median value.
		cre.ConsensusMedianAggregation[*big.Int](),
	)

	// Await the final, aggregated result.
	result, err := mathPromise.Await()
	if err != nil {
		return nil, err
	}

	logger.Info("Successfully fetched and aggregated math result", "result", result)

	return &MyResult{
		Result: result,
	}, nil
}

func main() {
	wasm.NewRunner(cre.ParseJSON[Config]).Run(InitWorkflow)
}
